src/Microsoft.ML.TimeSeries/Deseasonality/MeanDeseasonality.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace Microsoft.ML.TimeSeries.Deseasonality
+{
+    internal sealed class MeanDeseasonality : DeseasonalityBase
+    {
+        private double[] _circularComponent;
+
+        public override void Deseasonality(ref double[] values, int period, ref double[] results)
+        {
+            AllocateDoubleArray(period);
+            var length = values.Length;
+
+            // initialize the circurlar component to 0.
+            for (int i = 0; i < period; ++i)
+            {
+                _circularComponent[i] = 0;
+            }
+
+            // sum up values that locates at the same position in one period.
+            for (int i = 0; i < length; ++i)
+            {
+                var indexInPeriod = i % period;
+                _circularComponent[indexInPeriod] += values[i];
+            }
+
+            // calculate the mean value as circular component.
+            var cnt = (length - 1) / period;
+            var rest = (length - 1) % period;
+            for (int i = 0; i < period; ++i)
+            {
+                var lastCircle = i <= rest ? 1 : 0;
+                _circularComponent[i] = _circularComponent[i] / (cnt + lastCircle);
+            }
+
+            // substract the circular component from the original series.
+            for (int i = 0; i < length; ++i)
+            {
+                var indexInPeriod = i % period;
+                results[i] -= _circularComponent[indexInPeriod];
+            }
+        }
+
+        private void AllocateDoubleArray(int length)
+        {
+            if (_circularComponent == null)
+            {
+                _circularComponent = new double[length];
+            }
+            else if (_circularComponent.Length != length)
+            {
+                Array.Resize<double>(ref _circularComponent, length);
+            }
+        }
```

Comment:
I believe that Array.Resize<double>() already creates a new array if _circularComponent is `null`. So maybe there's no need to check if _circularComponent == null explicitly?

From https://docs.microsoft.com/en-us/dotnet/api/system.array.resize?view=netcore-3.1#remarks :
> If array is null, this method creates a new array with the specified size.

Reply:
The same applies to the other methods with the same name that you have in the other Deseasonality classes :wink:

---
In reply to: [440452024](https://github.com/dotnet/machinelearning/pull/5202#discussion_r440452024) [](ancestors = 440452024)

Reply:
In fact, since there's no need to check for `_circularComponent == null` then maybe you can get rid of this method altogether?

---
In reply to: [440455328](https://github.com/dotnet/machinelearning/pull/5202#discussion_r440455328) [](ancestors = 440455328,440452024)

Reply:
nice suggestion, updated

---
In reply to: [440502827](https://github.com/dotnet/machinelearning/pull/5202#discussion_r440502827) [](ancestors = 440502827,440455328,440452024)